crates/bevy_render/src/view/visibility/range.rs

```diff
@@ -0,0 +1,429 @@
+            }
+            _ => false,
+        }
+    }
+
+    /// Returns true if the entity is in range of any view.
+    ///
+    /// This only checks [`VisibilityRange`]s and doesn't perform any frustum or
+    /// occlusion culling. Thus the entity might not *actually* be visible.
+    ///
+    /// The entity is assumed to have a [`VisibilityRange`] component. If the
+    /// entity doesn't have that component, this method will return false.
+    #[inline]
+    pub fn entity_is_in_range_of_any_view(&self, entity: Entity) -> bool {
+        self.entities.contains_key(&entity)
+    }
+}
+
+/// Checks all entities against all views in order to determine which entities
+/// with [`VisibilityRange`]s are potentially visible.
+///
+/// This only checks distance from the camera and doesn't frustum or occlusion
+/// cull.
+pub fn check_visibility_ranges(
+    mut visible_entity_ranges: ResMut<VisibleEntityRanges>,
+    view_query: Query<(Entity, &GlobalTransform), With<Camera>>,
+    mut entity_query: Query<(Entity, &GlobalTransform, &VisibilityRange)>,
+) {
+    visible_entity_ranges.clear();
+
+    // Early out if the visibility range feature isn't in use.
+    if entity_query.is_empty() {
+        return;
+    }
+
+    // Assign an index to each view.
+    let mut views = vec![];
+    for (view, view_transform) in view_query.iter().take(32) {
+        let view_index = views.len() as u8;
+        visible_entity_ranges.views.insert(view, view_index);
+        views.push((view, view_transform.translation_vec3a()));
+    }
+
+    // Check each entity/view pair. Only consider entities with
+    // [`VisibilityRange`] components.
+    for (entity, entity_transform, visibility_range) in entity_query.iter_mut() {
+        let mut visibility = 0;
+        for (view_index, &(_, view_position)) in views.iter().enumerate() {
+            if visibility_range
+                .is_visible_at_all((view_position - entity_transform.translation_vec3a()).length())
+            {
+                visibility |= 1 << view_index;
+            }
+        }
+
+        // Invisible entities have no entry at all in the hash map. This speeds
+        // up checks slightly in this common case.
+        if visibility != 0 {
+            visible_entity_ranges.entities.insert(entity, visibility);
+        }
+    }
+}
+
+/// Extracts all [`VisibilityRange`] components from the main world to the
+/// render world and inserts them into [`RenderVisibilityRanges`].
+pub fn extract_visibility_ranges(
+    mut render_visibility_ranges: ResMut<RenderVisibilityRanges>,
+    visibility_ranges_query: Extract<Query<(Entity, &VisibilityRange)>>,
+    changed_ranges_query: Extract<Query<Entity, Changed<VisibilityRange>>>,
+) {
+    if changed_ranges_query.is_empty() {
+        return;
+    }
+
+    render_visibility_ranges.clear();
+    for (entity, visibility_range) in visibility_ranges_query.iter() {
+        render_visibility_ranges.insert(entity, visibility_range);
+    }
+}
```

Comment:
Why is the entire `RenderVisibilityRanges` cleared? Is it not possible to update only the changes ranges?

Reply:
If entities get removed, then `Changed` won't fire for them. So we'd need some more complex logic with `RemovedComponents` to make sure we handle that case and clear out removed entities from the list. This seemed like premature optimization.